### PR TITLE
events: Remove brittle `needsInitialFetch` system

### PR DESCRIPTION
**Pull request by chrisbobbe**
This is the next PR in my series toward better offline handling for the `/register` request.

We've long thought that the current logic is complex and fragile;
see for example
  https://chat.zulip.org/#narrow/stream/243-mobile-team/topic/Stuck.20on.20loading.20screen/near/907688

### CODE REVIEW

File: src/boot/StoreHydratedGate.js [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+      await dispatch(registerAndStartPolling());
+
+      // TODO(#3881): Lots of issues with outbox sending
+      dispatch(sendOutbox());
+
+      dispatch(initNotifications());
+    }
+  }, [dispatch, hasAuth]);
+
+  useEdgeTriggeredEffect(
```
Comment (gnprice) [inline src/boot/StoreHydratedGate.js:43]: In the final version, where this is in `StoreHydratedGate` and so `isHydrated` may start out false, I agree `useEdgeTriggeredEffect` is a good fit.

Continuing on from the comment above, though: I think the dependency on `hasAuth` should be taken as a genuine breach of the Rules of Hooks -- one that we're deliberately choosing here -- even though the ESLint plugin doesn't spot it. So there should be a comment explaining why that's OK. I think the point is basically that if `hasAuth` becomes true *later*, that's going to be either via ACCOUNT_SWITCH or LOGIN_SUCCESS, and each of those is accompanied by its own invocation of basically the same things as are in this `init` function.

File: src/events/eventActions.js [hunk 2]
```diff
@@ -320,8 +319,4 @@ export const deadQueue = (): ThunkAction<Promise<void>> => async (dispatch, getS
 
   // TODO(#3881): Lots of issues with outbox sending
   dispatch(sendOutbox());
-
-  // TODO: Remove this. No need to do this on DEAD_QUEUE or ACCOUNT_SWITCH:
-  //   https://chat.zulip.org/#narrow/stream/48-mobile/topic/.60remoteNotificationsRegistered.60.20listener.20too.20late/near/1148915
-  dispatch(initNotifications());
```
Comment (gnprice) [inline src/events/eventActions.js]: Remind me: was there something that actually goes wrong as a result of this, that removing it fixes? Or was it just a matter of feeling that this wasn't the right place for it?

As I think more about this action, I think it's actually in a situation very much parallel to `sendOutbox`: there's some data we have (be it our notifications token, or the user's messages they've tried to send); we want to get it to the server; when we eventually succeed at that, we'll update our data to reflect that, so we always have a to-do list of what we still want to send; and we want to keep trying in order to get it through to the server, until we succeed (or in the case of a message, perhaps time out.)

Those, and: we don't currently have a real principled way of driving those retries. Instead, we take the occasion of the initial fetch as a hook to get us to retry. It'd be good to get a principled way of doing them (for outbox, that's #3881), but until then every place we retry is potentially a helpful one, even if ad hoc.

So if there isn't a concrete problem that's solved by removing these `initNotification` calls, I think I'd prefer to actually keep them. After #3881, we can take whatever we work out for that and make something similar for driving these retries too.

**Review by gnprice — COMMENTED** [review]
Thanks @chrisbobbe! This code around `needsInitialFetch` has been a tangly part of our control flow -- very glad to see it cleaned up. Comments below.

File: src/events/eventActions.js [hunk 3]
```diff
@@ -3,22 +3,197 @@ import { addBreadcrumb } from '@sentry/react-native';
...
+ * up, or regaining a network connection. We fetch private messages
+ * here so that we can show something useful in the PM conversations
+ * screen, but we hope to stop doing this soon (see note at
+ * `fetchPrivateMessages`). We fetch messages in a few other places:
+ * to initially populate a message list (`ChatScreen`), to grab more
+ * messages on scrolling to the top or bottom of the message list
+ * (`fetchOlder` and `fetchNewer`), and to grab search results
+ * (`SearchMessagesScreen`).
+ */
+export const registerAndStartPolling = (): ThunkAction<Promise<void>> => async (
```
[the remark this answers is not in the document]
Comment (chrisbobbe) [reply src/events/eventActions.js:113]: Ah, good thought. Here's an important place where it shows up: https://zulip.readthedocs.io/en/latest/subsystems/events-system.html#the-initial-data-fetch

File: src/account/accountActions.js [hunk 4]
```diff
@@ -24,9 +32,37 @@ const accountSwitchPlain = (index: number): AllAccountsAction => ({
...
+     - The PerAccountState that Dispatch currently acts on is the one
+       belonging to the active account.
+     - We want this dispatch to act on the active account -- the new,
+       post-switch active account.
+     - It will act on that account, because at this point we've already
+       dispatched `accountSwitchPlain`.
+
+     TODO(#5006): maybe have a separate "active account" thunk action type?
+  */
+  const activeAccountDispatch: Dispatch = dispatch;
```
[the remark this answers is not in the document]
Comment (chrisbobbe) [reply src/account/accountActions.js:60]: > But what it means to refer to is an "extra" in the redux-thunk-action sense. We have several of those now for per-account actions, described here: […]

Oh! Got it; this is making more sense now, thanks. 🙂

File: src/boot/StoreHydratedGate.js [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+    () => {
+      init();
+    },
```
[the remark this answers is not in the document]
Comment (chrisbobbe) [reply src/boot/StoreHydratedGate.js:46]: `init` returns a Promise.

`useEdgeTriggeredEffect`, as well as `useEffect` and our new `useConditionalEffect` we discussed in the office on Thursday, won't take a callback that returns a Promise.

In fact, if I just pass an async function expression directly to `useEffect`—

```js
useEffect(async () => 'foo');
```

—I get a complaint from `react-hooks/exhaustive-deps`:

```
Effect callbacks are synchronous to prevent race conditions. Put the async function inside:

useEffect(() => {
  async function fetchData() {
    // You can await here
    const response = await MyAPI.getData(someId);
    // ...
  }
  fetchData();
}, [someId]); // Or [] if effect doesn't need props or state

Learn more about data fetching with Hooks: https://reactjs.org/link/hooks-data-fetchingeslint[react-hooks/exhaustive-deps](https://github.com/facebook/react/issues/14920)
```

File: src/events/eventActions.js (hunk 2 above)
Comment (chrisbobbe) [reply src/events/eventActions.js]: > Remind me: was there something that actually goes wrong as a result of this, that removing it fixes? Or was it just a matter of feeling that this wasn't the right place for it?

Yeah, as far as I remember it's the latter. I'll adjust.

Comment (chrisbobbe) [reply src/events/eventActions.js]: Yeah, we'll have to think carefully about retries.

Even setting that aside, I've often found our code for registering push tokens pretty hard to follow, and wondered if we could make it better.

We quick-fixed a P0 bug #4573 with 6ffc7e18a, and I don't think we've settled on a plan for something more permanent.

I think this is the description of what the relevant subsystem is aimed at:

For all servers we want notifications from, we should keep the server up-to-date with our current device token.
- If we're logged into a server, we want notifications from it.
  - This suggests having some per-account code that runs with the account's LOGIN_SUCCESS.
- Our device token could change between sessions of running the app, or during a session.
  - This suggests having some global code that runs on startup, and on learning (e.g., from an event) that the token has changed during this session. Global because the token is for the device, and also because we may need to tell multiple servers about the token.

I think there's a main reason why our code has developed such that it's not super clear how we answer these requirements: we're not supposed to ask iOS for the token until we're ready to pop up a permissions request, and we're supposed to do that only in a context where the permission would be useful to have. We've (reasonably) decided that the app enters that context when a `/register` completes, so that's when we request the token on iOS.

Here's an example of how the code is hard to follow. It occurred to me after your comment prompted me to keep thinking about this:

> - Our device token could change between sessions of running the app, or during a session.
>   - This suggests having some global code that runs on startup […]. Global because the token is for the device, and also because we may need to tell multiple servers about the token.

Our "on-startup" code for this is `initNotifications`. That sure doesn't look like "some global code", does it? It's a `ThunkAction`, not a `GlobalThunkAction`. But then, if you sit and think about it, and read some distant code, it's possible to convince yourself that all interested servers will indeed be notified if the device token changed between sessions of running the app. Better if the "on-startup" code were instead a `GlobalThunkAction`, with a nice jsdoc that mentioned that device tokens can change between sessions.

[I have a draft](https://chat.zulip.org/#narrow/stream/48-mobile/topic/.60remoteNotificationsRegistered.60.20listener.20too.20late/near/1147424) from a year ago for a unified `getAndReturnNotificationToken` function that grabs the token in the platform-appropriate way. That specific draft may or may not end up being helpful, but I think if we chose one of the following, we could unify our separate iOS/Android code and thus remove lots of complexity:

1. iOS and Android both get the token immediately on startup (sad for iOS because the permission pop-up isn't deferred till after `/register`), or
2. iOS and Android both get the token after `/register`, where iOS gets it currently (sad for Android because this isn't necessary, but also it isn't harmful, is it?)

**Comment by chrisbobbe** [comment]
Thanks for the review! Revision pushed.

File: src/boot/AppDataFetcher.js [hunk 6]
```diff
@@ -1,25 +1,47 @@
...
-  }, [needsInitialFetch, dispatch]);
+  }, [dispatch, hasAuth]);
+
+  useEffect(
+    () => {
+      init();
+    },
+    // We just want to init once, just after the store rehydrates. This
+    // component will have begun its life at that time, since it's under
+    // HideIfNotHydrated.
```
Comment (gnprice) [inline src/boot/AppDataFetcher.js:40]: nit:
```suggestion
    // StoreHydratedGate.
```

File: src/boot/StoreHydratedGate.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (gnprice) [reply src/boot/StoreHydratedGate.js]: I see, yeah. Here we're happy for it to be an async thing we don't wait for, but it is good to have something making that explicit. Which this does, a little bit.

File: src/events/eventActions.js [hunk 7]
```diff
@@ -129,3 +304,19 @@ export const startEventPolling = (
     lastEventId = Math.max(lastEventId, ...events.map(x => x.id));
   }
 };
+
+const deadQueuePlain = (): PerAccountAction => ({
+  type: DEAD_QUEUE,
+});
+
+export const deadQueue = (): ThunkAction<Promise<void>> => async (dispatch, getState) => {
```
[the remark this answers is not in the document]
Comment (gnprice) [reply src/events/eventActions.js:316]: This can still be an unexported `const deadQueue` when it becomes a thunk action.

File: src/boot/AppDataFetcher.js [hunk 8]
```diff
@@ -1,25 +1,47 @@
...
+    //
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+    [],
```
Comment (gnprice) [inline src/boot/AppDataFetcher.js:43]: I think this still calls for an explanation of why it's OK that we only call this once, even if `hasAuth` later changes. Along the lines of here:
https://github.com/zulip/zulip-mobile/pull/5300#discussion_r829577313

File: src/events/eventActions.js (hunk 2 above)
Comment (gnprice) [reply src/events/eventActions.js]: > I think this is the description of what the relevant subsystem is aimed at:

Yeah, that all seems right.



> Here's an example of how the code is hard to follow. It occurred to me after your comment prompted me to keep thinking about this: […]
> 
> Our "on-startup" code for this is `initNotifications`. That sure doesn't look like "some global code", does it? It's a `ThunkAction`, not a `GlobalThunkAction`. But then, if you sit and think about it and read some distant code, it's possible to convince yourself that all interested servers will indeed be notified if the device token changed between sessions of running the app, on iOS and Android.

Hmm, indeed. I agree, that seems like a key point for why this logic is hard to follow.

Want to make an issue for this? That'll probably provide a better way to track it than this review subthread.

**Comment by gnprice** [comment]
Thanks! This generally looks good; small comments above.

File: src/events/eventActions.js (hunk 2 above)
Comment (chrisbobbe) [reply src/events/eventActions.js]: Sure, sounds good: #5329.

**Comment by chrisbobbe** [comment]
Thanks for the review! Revision pushed.

**Comment by chrisbobbe** [comment]
Just rebased (whew, there was a lot to do this time, especially after #5393!)

**Comment by gnprice** [comment]
> Just rebased (whew, there was a lot to do this time, especially after #5393 [the Prettier 2 upgrade]!)

Hmm, yeah -- we didn't discuss workflow for open PRs, did we. For comparable changes in zulip/zulip, there's generally an explicit set of steps pre-developed, tested, and then announced to contributors, for how to rebase their work past the change.

I think the basic algorithm that should reliably work, and avoids a lot of manual conflict resolution, is:
* Rebase onto the commit just before the formatting change. Resolve conflicts along the way as usual.
* Rebase onto the commit that made the formatting change. Resolve each conflict mechanically, by taking your changed code and auto-formatting it: like `git checkout --theirs . && tools/test prettier --fix --diff @ && git add -u && git rebase --continue`. (Well, and prefixed with `yarn install &&` the first time.)
* Rebase onto the latest upstream. Resolve conflicts along the way as usual.

I did something like that for one complex PR I had open at the time, and it worked well. I have plenty of draft branches I'll probably have to do something similar with, and you likely do too. I guess let's discuss in a chat thread with whatever we find the next time one of us does a rebase that calls for this. Should be possible to distill it into a short copy-pasteable script, and/or a little script in `tools/` (though it's a bit tricky to use the latter, since by definition one's acting on old branches.)

**Comment by chrisbobbe** [comment]
Oh, yeah, that's a very good idea. 🙂

**Review by gnprice — COMMENTED** [review]
OK, revision reviewed! I also re-read the whole branch to refresh myself on all the changes and make one more review pass on it, given that this is reworking such a core subsystem.

Everything looks good except a few small comments below -- all of which are on the three newer commits at the end. Those are close, but this PR thread has already gotten long for GitHub. So I think a good plan is:
* I'll merge the other 15 commits. :tada: 
* Then please send a revision of those last 3 commits as a fresh PR.
